File: clients/typescript/src/models/agent-profile.ts
/**
 * AgentProfile — TypeScript mirror of the Python SDK's
 * `openhands.sdk.profiles.agent_profile` module.
 *
 * The AgentProfile union is discriminated on `agent_kind` and carries
 * *references* (llm_profile_ref / mcp_server_refs) rather than resolved
 * credentials, mirroring the Python design. See epic #3713.
 */

import type { ACPProviderKey } from './acp';

// ── Shared supporting types ──────────────────────────────────────────────────

/** Discriminator across the {@link AgentProfile} union. */
export type AgentKind = 'openhands' | 'acp';

/** ACP backend selector: a built-in provider key or the `'custom'` sentinel. */
export type ACPServerKind = ACPProviderKey | 'custom';

/** Secret-free critic/refinement policy stored inside an OpenHands profile. */
export interface ProfileVerificationSettings {
  critic_enabled: boolean;
  critic_mode: string;
  enable_iterative_refinement: boolean;
  critic_threshold: number;
  max_refinement_iterations: number;
  critic_server_url: string | null;
  critic_model_name: string | null;
}

// ── Profile variants ─────────────────────────────────────────────────────────

interface AgentProfileBase {
  schema_version?: number;
  /** Stable UUID handle — never changes on rename. */
  id: string;
  /** Human-facing, renameable key. */
  name: string;
  /** Monotonic revision counter, bumped on each saved edit. */
  revision: number;
  /**
   * Which of the user's globally configured MCP servers to expose.
   * `null` = all; `[]` = none; a non-null list = filter to the named keys.
   */
  mcp_server_refs: string[] | null;
  /**
   * Conversation secret allow-list. `null` preserves supplied secrets without
   * loading additional saved secrets; `[]` exposes none. A list selects only
   * those names, resolving matching saved secrets at launch. ACP profiles must
   * include their provider credential when using an explicit list.
   */
  secret_refs?: string[] | null;
}

/** `agent_kind="openhands"` variant — references an LLM profile by name. */
export interface OpenHandsAgentProfile extends AgentProfileBase {
  agent_kind: 'openhands';
  /** Name of the LLM profile to resolve. */
  llm_profile_ref: string;
  agent: string;
  skills: unknown[];
  system_message_suffix: string | null;
  condenser: unknown;
  verification: ProfileVerificationSettings;
  enable_sub_agents: boolean;
  tool_concurrency_limit: number;
}

/** `agent_kind="acp"` variant — names an ACP backend; stores no credential. */
export interface ACPAgentProfile extends AgentProfileBase {
  agent_kind: 'acp';
  /** ACP provider key, or `'custom'` for a user-supplied command. */
  acp_server: ACPServerKind;
  acp_model: string | null;
  acp_session_mode: string | null;
  acp_prompt_timeout: number;
  acp_command: string | null;
  acp_args: string[] | null;
}

/**
 * Discriminated union on `agent_kind`. Use the `agent_kind` field to narrow:
 * ```ts
 * if (profile.agent_kind === 'openhands') { ... }
 * ```
 */
export type AgentProfile = OpenHandsAgentProfile | ACPAgentProfile;

/**
 * Request body for `POST /api/agent-profiles/{name}` (create or overwrite).
 *
 * Discriminated on `agent_kind`, so a payload cannot mix fields from the other
 * variant (the server model is `extra=forbid`). Non-discriminant fields are
 * optional: server-managed identity (`id`/`revision`/`schema_version`) is
 * minted on create and preserved on overwrite. A field left unset takes the
 * server-side default where one exists — except `llm_profile_ref`, which is
 * required for the `openhands` variant (omitting it yields a 422).
 */
export type AgentProfileSaveInput =
  | ({ agent_kind: 'openhands' } & Partial<Omit<OpenHandsAgentProfile, 'agent_kind'>>)
  | ({ agent_kind: 'acp' } & Partial<Omit<ACPAgentProfile, 'agent_kind'>>);

// ── Summary projection (list endpoint) ──────────────────────────────────────

/** Lightweight summary returned by the list endpoint. */
export interface AgentProfileSummary {
  id: string | null;
  name: string;
  agent_kind: AgentKind;
  revision: number | null;
  llm_profile_ref: string | null;
  mcp_server_refs: string[] | null;
}

// ── Materialize diagnostics ──────────────────────────────────────────────────

/**
 * Dry-run resolve report returned by `POST /api/agent-profiles/{name}/materialize`.
 * Mirrors `openhands.sdk.profiles.resolver.AgentProfileDiagnostics` field-for-field.
 *
 * Dangling references are reported here (valid=false) rather than as HTTP errors.
 */
export interface AgentProfileDiagnostics {
  agent_kind: AgentKind;
  valid: boolean;
  errors: string[];

  /** Secret scope; `null` = unrestricted. */
  secret_refs?: string[] | null;

  // OpenHands LLM reference.
  llm_profile_ref: string | null;
  llm_profile_resolved: boolean;
  llm_api_key_set: boolean;

  // MCP composition (both variants).
  mcp_server_refs: string[] | null;
  resolved_mcp_servers: string[];
  dangling_mcp_server_refs: string[];

  // ACP provider credential channels (ACP variant only).
  acp_api_key_secret_name: string | null;
  acp_base_url_secret_name: string | null;
  acp_file_secret_names: string[];

  // Redacted resolved settings, present iff valid.
  resolved_settings: Record<string, unknown> | null;
}

// ── Launch provenance ──────────────────────────────────────────────────────

/** Mirrors `openhands.sdk.profiles.agent_profile.LaunchedAgentProfile`. */
export interface LaunchedAgentProfile {
  agent_profile_id: string;
  revision: number;
  /** Launch-time secret allow-list, also enforced on resume. */
  secret_refs?: string[] | null;
}

/**
 * Provenance snapshot recorded when a profile launches a conversation.
 * @deprecated Use LaunchedAgentProfile, which mirrors the server payload.
 *
 * Stored on the conversation so clients can identify which profile is current
 * without fragile settings-comparison. See epic #3713.
 */
export interface LaunchedProfile {
  /** Stable id of the profile that launched the conversation. */
  profile_id: string;
  /** Revision of the profile at launch time. */
  revision: number;
}

File: openhands-sdk/openhands/sdk/profiles/agent_profile.py
"""``AgentProfile`` — the named, reference-bearing agent launch spec.

A separate ``Discriminator("agent_kind")`` + ``Tag`` union from
:data:`~openhands.sdk.settings.model.AgentSettingsConfig`: the profile carries
*references* (``llm_profile_ref`` / ``mcp_server_refs``) and is secret-free at
rest, whereas the settings union embeds the resolved ``llm`` / ``mcp_config``.
See epic #3713 for the resolution model.
"""

from __future__ import annotations

from collections.abc import Callable, Mapping
from typing import Annotated, Any, Literal
from uuid import UUID, uuid4

from pydantic import (
    BaseModel,
    ConfigDict,
    Discriminator,
    Field,
    Tag,
    TypeAdapter,
    ValidationError,
)

from openhands.sdk.settings.model import (
    ACPServerKind,
    CondenserSettingsConfig,
    CriticMode,
    LLMSummarizingCondenserSettings,
    VerificationSettings,
)
from openhands.sdk.tool import Tool


AGENT_PROFILE_SCHEMA_VERSION = 2


class ProfileVerificationSettings(BaseModel):
    """Secret-free critic/refinement policy for a profile.

    The non-credential subset of
    :class:`~openhands.sdk.settings.model.VerificationSettings`: ``critic_api_key``
    is omitted so the profile holds no secret. The critic reuses the resolved
    LLM profile's key (the existing ``critic_api_key=None`` behavior).
    """

    critic_enabled: bool = False
    critic_mode: CriticMode = "finish_and_message"
    enable_iterative_refinement: bool = False
    critic_threshold: float = Field(default=0.6, ge=0.0, le=1.0)
    max_refinement_iterations: int = Field(default=3, ge=1)
    critic_server_url: str | None = None
    critic_model_name: str | None = None


def build_profile_verification(
    v: VerificationSettings,
) -> ProfileVerificationSettings:
    """Project the secret-free subset of ``VerificationSettings`` onto a profile.

    Drops ``critic_api_key`` — the profile is secret-free; the critic reuses the
    resolved LLM profile's key. Hoisted from the agent-server router so the
    local and cloud seeds build verification identically.
    """
    return ProfileVerificationSettings(
        critic_enabled=v.critic_enabled,
        critic_mode=v.critic_mode,
        enable_iterative_refinement=v.enable_iterative_refinement,
        critic_threshold=v.critic_threshold,
        max_refinement_iterations=v.max_refinement_iterations,
        critic_server_url=v.critic_server_url,
        critic_model_name=v.critic_model_name,
    )


class AgentProfileBase(BaseModel):
    """Shared identity + provenance fields for every ``AgentProfile`` variant.

    ``extra="forbid"`` is what gives the union its cross-variant safety: a
    payload tagged ``agent_kind="acp"`` that also carries ``llm_profile_ref``
    (or an OpenHands payload carrying ``acp_*``) is rejected rather than
    silently dropping the foreign field into a mongrel profile.
    """

    model_config = ConfigDict(extra="forbid")

    schema_version: int = Field(default=AGENT_PROFILE_SCHEMA_VERSION, ge=1)
    id: UUID = Field(
        default_factory=uuid4,
        description=(
            "Stable provenance handle for this profile. Conversations record "
            "this UUID; it never changes, even when the profile is renamed."
        ),
    )
    name: str = Field(
        min_length=1,
        description="Human-facing, renameable key shown in the profile picker.",
    )
    revision: int = Field(
        default=0,
        ge=0,
        description="Monotonic revision counter, bumped on each saved edit.",
    )
    # null = all of the user's MCP servers; [] = none; a non-null list filters
    # to the named keys. null and [] are deliberately distinct. MCP servers are
    # user-configured and persisted, so this reference is safe: every name that
    # can be referenced is present in ``mcp_config``. Skills are NOT modeled this
    # way — they are discovered from many lazy/incomplete sources, so a profile
    # selects them by *exclusion* (``OpenHandsAgentProfile.disabled_skills``, a
    # deny-list) rather than by an allow-list of names that can dangle. See the
    # #4017 architecture discussion.
    mcp_server_refs: list[str] | None = Field(
        default=None,
        description=(
            "Which of the user's globally configured MCP servers to expose. "
            "null = all; [] = none; a non-null list = filter to the named keys."
        ),
    )
    # Names only; selected saved values are resolved when launching a conversation.
    secret_refs: list[str] | None = Field(
        default=None,
        description=(
            "Conversation secret allow-list. null preserves supplied secrets without "
            "loading additional saved secrets; [] exposes none. A list selects only "
            "those names, resolving matching saved secrets at launch. An ACP profile "
            "must include its provider credential when using an explicit list."
        ),
    )


class OpenHandsAgentProfile(AgentProfileBase):
    """``agent_kind="openhands"`` profile — references an LLM profile by name.

    Mirrors the configurable surface of
    :class:`~openhands.sdk.settings.model.OpenHandsAgentSettings`, except the
    concrete ``llm`` is replaced by :attr:`llm_profile_ref` (resolved against
    the LLM profile store) and ``mcp_config`` by the inherited
    :attr:`~AgentProfileBase.mcp_server_refs`.
    """

    agent_kind: Literal["openhands"] = Field(
        default="openhands",
        description=(
            "Discriminator for the ``AgentProfile`` union. ``'openhands'`` "
            "selects the standard built-in OpenHands agent."
        ),
    )
    llm_profile_ref: str = Field(
        min_length=1,
        description=(
            "Name of the saved LLM profile to resolve for this agent. The "
            "profile itself stores no LLM credential — it lives on the "
            "referenced LLM profile."
        ),
    )
    agent: str = Field(
        default="CodeActAgent",
        description="Agent class to build.",
    )
    # Same tri-state as the resolved settings' ``tools``: passed through
    # verbatim by the resolver, so ``create_agent`` is the single defaulting
    # point (#3978). Secret-free by construction (``Tool`` is name + params).
    tools: list[Tool] | None = Field(
        default=None,
        description=(
            "Tool selection for the resolved agent. None (the default) = the "
            "server's standard tool set; [] = an explicitly bare agent; a "
            "non-empty list is used exactly as given."
        ),
    )

    system_message_suffix: str | None = Field(
        default=None,
        description="Optional suffix appended to the system prompt.",
    )
    disabled_skills: list[str] = Field(
        default_factory=list,
        description=(
            "Names of server-discovered skills to EXCLUDE from this agent's "
            "prompt — a deny-list over the discovered catalog. [] (the default) "
            "= all discovered skills; a listed name that is absent from the "
            "catalog is simply a no-op, so the field never fails a launch when "
            "the catalog drifts. Replaces the former allow-list ``skill_refs`` "
            "(#4017): skills are discovered from many incomplete sources, so "
            "exclusion is the only selection model that can't dangle."
        ),
    )
    condenser: CondenserSettingsConfig = Field(
        default_factory=LLMSummarizingCondenserSettings,
        description="Condenser settings for the agent.",
    )
    verification: ProfileVerificationSettings = Field(
        default_factory=ProfileVerificationSettings,
        description="Critic/verification policy (secret-free; no critic_api_key).",
    )
    enable_sub_agents: bool = Field(
        default=False,
        description="Enable sub-agent delegation via TaskToolSet.",
    )
    enable_switch_llm_tool: bool = Field(
        default=True,
        description=(
            "Enable the built-in switch_llm tool for switching between saved "
            "LLM profiles. Defaults True to match the global agent settings "
            "default (AgentSettingsConfig.enable_switch_llm_tool)."
        ),
    )
    tool_concurrency_limit: int = Field(
        default=1,
        ge=1,
        description=(
            "Maximum number of tool calls to execute concurrently per agent "
            "step. 1 = sequential (default)."
        ),
    )


class ACPAgentProfile(AgentProfileBase):
    """``agent_kind="acp"`` profile — names an ACP backend, stores no credential.

    There is no ``llm_profile_ref`` and no embedded credential: the ACP
    subprocess makes its own model calls, and the provider's credential
    (env-var name) is derived from :attr:`acp_server` via the
    :data:`~openhands.sdk.settings.acp_providers.ACP_PROVIDERS` registry. The
    value rides the conversation secrets channel, never the profile.
    """

    agent_kind: Literal["acp"] = Field(
        default="acp",
        description=(
            "Discriminator for the ``AgentProfile`` union. ``'acp'`` selects an "
            "ACP-delegating agent."
        ),
    )
    # No skill-selection field: ACP agents own their tooling and prompt
    # construction. Project skills never reach one (the CLI reads the repo
    # itself, #4019); whether any managed skill does is a deployment choice the
    # caller expresses through ``resolve_agent_profile``'s ``available_skills``.
    acp_server: ACPServerKind = Field(
        default="claude-code",
        description=(
            "Which ACP-compatible backend to launch. The provider's credential "
            "env-var name is derived from this via the ACP_PROVIDERS registry."
        ),
    )
    acp_model: str | None = Field(
        default=None,
        description=(
            "Model identifier for the ACP server (e.g. 'claude-opus-4-8'). "
            "Leave blank to let the server pick its default."
        ),
    )
    acp_session_mode: str | None = Field(
        default=None,
        description=(
            "Session mode ID (e.g. 'bypassPermissions'). Leave blank to "
            "auto-detect from the ACP server type."
        ),
    )
    acp_prompt_timeout: float = Field(
        default=1800.0,
        gt=0,
        description=(
            "Inactivity timeout (seconds) for a single ACP prompt() round-trip; "
            "resets on every update from the server."
        ),
    )
    acp_startup_timeout: float = Field(
        default=90.0,
        gt=0,
        description=(
            "Timeout (seconds) for ACP server startup: spawn, "
            "initialize/authenticate, and new_session()/load_session()."
        ),
    )
    acp_command: str | None = Field(
        default=None,
        description=(
            "Optional explicit command to launch the ACP subprocess. Leave "
            "blank to use the default for ``acp_server``."
        ),
    )
    acp_args: list[str] | None = Field(
        default=None,
        description="Additional arguments appended to the ACP server command.",
    )


class LaunchedAgentProfile(BaseModel):
    """Provenance snapshot recorded when an agent profile launches a conversation.

    Stored on ``StoredConversation`` and projected onto ``ConversationInfo`` so
    ts-client ``deriveSwitchPlan`` can identify which agent profile is current
    without fragile settings-comparison. See #3720.
    """

    agent_profile_id: UUID = Field(
        description="Stable id of the agent profile that launched the conversation.",
    )
    revision: int = Field(
        ge=0,
        description="Revision of the agent profile at launch time.",
    )
    secret_refs: list[str] | None = Field(
        default=None,
        description=(
            "Secret allow-list captured at launch, also enforced on resume. "
            "null preserves unrestricted behavior for older conversations."
        ),
    )

    def allows_secret(self, name: str) -> bool:
        return self.secret_refs is None or name in self.secret_refs


def _agent_profile_discriminator(value: Any) -> str:
    """Discriminator for :data:`AgentProfile` — defaults to ``'openhands'``.

    A payload without an explicit ``agent_kind`` is treated as the standard
    OpenHands variant, mirroring
    :func:`~openhands.sdk.settings.model._agent_settings_discriminator`.
    """
    if isinstance(value, BaseModel):
        return getattr(value, "agent_kind", "openhands")
    if isinstance(value, Mapping):
        return value.get("agent_kind", "openhands")
    return "openhands"


AgentProfile = Annotated[
    Annotated[OpenHandsAgentProfile, Tag("openhands")]
    | Annotated[ACPAgentProfile, Tag("acp")],
    Discriminator(_agent_profile_discriminator),
]
"""Discriminated union over the agent-profile variants.

Use :func:`validate_agent_profile` (or a :class:`~pydantic.TypeAdapter`) to
validate/construct instances from raw payloads.
"""


PersistedProfileMigrator = Callable[[dict[str, Any]], dict[str, Any]]


def _migrate_v1_to_v2(payload: dict[str, Any]) -> dict[str, Any]:
    """Remove the retired embedded-skills configuration from v1 profiles."""
    migrated = dict(payload)
    # ``skills`` was removed from the persisted profile shape in v2.  It must
    # be discarded here, before strict model validation, so a v1 profile can
    # be opened and saved as its canonical v2 representation.
    migrated.pop("skills", None)
    if (
        migrated.get("agent_kind", "openhands") == "openhands"
        and migrated.get("name") == "default"
        and migrated.get("revision", 0) == 0
        and migrated.get("tools") == []
    ):
        migrated["tools"] = None
    migrated["schema_version"] = 2
    return migrated


_AGENT_PROFILE_MIGRATIONS: dict[int, PersistedProfileMigrator] = {
    1: _migrate_v1_to_v2,
}


def _apply_persisted_migrations(payload: dict[str, Any]) -> dict[str, Any]:
    """Bring an ``AgentProfile`` payload up to the current schema."""
    migrated = dict(payload)
    version_raw = migrated.get("schema_version")
    if version_raw is None:
        migrated["schema_version"] = AGENT_PROFILE_SCHEMA_VERSION
        version = AGENT_PROFILE_SCHEMA_VERSION
    elif isinstance(version_raw, int) and not isinstance(version_raw, bool):
        version = version_raw
    else:
        raise TypeError(
            "AgentProfile schema_version must be an integer, got "
            f"{type(version_raw).__name__}."
        )

    if version < 0:
        raise ValueError("AgentProfile schema_version must be non-negative.")
    if version > AGENT_PROFILE_SCHEMA_VERSION:
        raise ValueError(
            f"AgentProfile schema_version {version} is newer than supported "
            f"version {AGENT_PROFILE_SCHEMA_VERSION}."
        )

    while version < AGENT_PROFILE_SCHEMA_VERSION:
        migrate = _AGENT_PROFILE_MIGRATIONS.get(version)
        if migrate is None:
            raise ValueError(
                f"No migration registered for AgentProfile schema_version {version}."
            )
        migrated = migrate(dict(migrated))
        next_version = migrated.get("schema_version")
        if not isinstance(next_version, int) or isinstance(next_version, bool):
            raise ValueError(
                f"Migration for AgentProfile schema_version {version} did not "
                "produce a valid integer schema_version."
            )
        if next_version <= version:
            raise ValueError(
                f"Migration for AgentProfile schema_version {version} did not "
                "advance the schema_version."
            )
        version = next_version

    return migrated


_AGENT_PROFILE_ADAPTER: TypeAdapter[OpenHandsAgentProfile | ACPAgentProfile] = (
    TypeAdapter(AgentProfile)
)


def validate_agent_profile(
    data: Any,
    *,
    context: Mapping[str, Any] | None = None,
) -> OpenHandsAgentProfile | ACPAgentProfile:
    """Load and validate an ``AgentProfile`` payload, narrowing on ``agent_kind``.

    Already-validated instances pass through unchanged. Raw mappings are
    migrated to the current schema version, then validated against
    :data:`AgentProfile`, so the return is always a canonical variant.
    """
    if isinstance(data, OpenHandsAgentProfile | ACPAgentProfile):
        return data
    if isinstance(data, BaseModel):
        payload = data.model_dump(mode="json")
    elif isinstance(data, Mapping):
        payload = dict(data)
    else:
        raise TypeError("AgentProfile payload must be a mapping or BaseModel.")
    payload = _apply_persisted_migrations(payload)
    return _AGENT_PROFILE_ADAPTER.validate_python(payload, context=context)


def safe_validation_error_detail(exc: ValidationError) -> list[dict[str, Any]]:
    """FastAPI-shaped ``detail`` for a 422 from a failed profile validation.

    Surfaces ``loc``/``type``/``msg`` per error. The profile carries no
    secret-bearing field — embedded ``skills[].mcp_tools`` was removed in #4017,
    and the ``disabled_skills`` deny-list holds only skill *names* — so the full
    validation message is safe to return, and a client needs it to see *why* a
    save failed (e.g. an invalid ``acp_server`` value). ``input`` is still
    dropped: it echoes the whole rejected payload, which is noisy and not needed
    to explain the error. Shaped like FastAPI's request-validation ``detail`` (a
    list of error objects) so routers can hand it straight to ``HTTPException``.
    Hoisted from the agent-server router so the local and cloud routers behave
    identically.
    """
    return [
        {"loc": list(err["loc"]), "type": str(err["type"]), "msg": str(err["msg"])}
        for err in exc.errors()
    ]
